stratum/hal/lib/barefoot/p4runtime_bfrt_translator.cc

```diff
@@ -0,0 +1,742 @@
+            break;
+          }
+          case ::p4::v1::FieldMatch::kRange: {
+            // Only accept "exact match" range value, which means both low
+            // and high value must be the same.
+            CHECK_RETURN_IF_FALSE(field_match.range().low() ==
+                                  field_match.range().high());
+            ASSIGN_OR_RETURN(const std::string& new_val,
+                             TranslateValue(field_match.range().low(), *uri,
+                                            to_sdk, to_bit_width));
+            field_match.mutable_range()->set_low(new_val);
+            field_match.mutable_range()->set_high(new_val);
+            break;
+          }
+          case ::p4::v1::FieldMatch::kOptional: {
+            ASSIGN_OR_RETURN(const std::string& new_val,
+                             TranslateValue(field_match.optional().value(),
+                                            *uri, to_sdk, to_bit_width));
+            field_match.mutable_optional()->set_value(new_val);
+            break;
+          }
+          default:
+            return MAKE_ERROR(ERR_UNIMPLEMENTED)
+                   << "Unsupported field match type: "
+                   << field_match.ShortDebugString();
+        }
+      }  // else, we don't modify the value if it doesn't need to be translated.
+    }
+  }
+
+  if (translated_entry.action().type_case() == ::p4::v1::TableAction::kAction) {
+    ASSIGN_OR_RETURN(
+        *(translated_entry.mutable_action()->mutable_action()),
+        TranslateAction(translated_entry.action().action(), to_sdk));
+  } else if (translated_entry.action().type_case() ==
+             ::p4::v1::TableAction::kActionProfileActionSet) {
+    auto* action_set =
+        translated_entry.mutable_action()->mutable_action_profile_action_set();
+    for (::p4::v1::ActionProfileAction& action_profile_action :
+         *action_set->mutable_action_profile_actions()) {
+      ASSIGN_OR_RETURN(*(action_profile_action.mutable_action()),
+                       TranslateAction(action_profile_action.action(), to_sdk));
+    }
+  }  // else, we don't translate action profile member id or group id.
+
+  return translated_entry;
+}
+
+::util::StatusOr<::p4::v1::ActionProfileMember>
+P4RuntimeBfrtTranslator::TranslateActionProfileMember(
+    const ::p4::v1::ActionProfileMember& act_prof_mem, bool to_sdk) {
+  ::p4::v1::ActionProfileMember translated_apm;
+  translated_apm.CopyFrom(act_prof_mem);
+  const auto& action_profile_id = act_prof_mem.action_profile_id();
+  const auto& member_id = act_prof_mem.member_id();
+  ASSIGN_OR_RETURN(*(translated_apm.mutable_action()),
+                   TranslateAction(translated_apm.action(), to_sdk));
+  return translated_apm;
+}
+
+::util::StatusOr<::p4::v1::MeterEntry>
+P4RuntimeBfrtTranslator::TranslateMeterEntry(const ::p4::v1::MeterEntry& entry,
+                                             bool to_sdk) {
+  // TODO(Yi Tseng): Will support this in another PR.
```

Comment:
I don't think there is anything to translate for indirect meters. Or just `TranslateIndex`?

Reply:
Right, this is an easy task. I include the support and test for Meter, DirectMeter, and Register in the PR